java/core/src/main/java/com/whylogs/core/ColumnProfile.java

```diff
@@ -0,0 +1,91 @@
+package com.whylogs.core;
+
+import com.whylogs.core.metrics.Metric;
+import com.whylogs.core.metrics.OperationResult;
+import com.whylogs.core.schemas.ColumnSchema;
+import com.whylogs.core.views.ColumnProfileView;
+import java.util.ArrayList;
+import java.util.HashMap;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+@Getter
+@ToString
+@EqualsAndHashCode
+public class ColumnProfile<T> implements AutoCloseable {
+  // Required
+  private final String name;
+  private final ColumnSchema schema;
+  private final int cacheSize;
+
+  // Has Defaults
+  private HashMap<String, Metric<?>> metrics;
+  private final SingleFieldProjector<T> projector;
+  private int successCount;
+  private int failureCount;
+  private int nullCount;
+
+  private ArrayList<T> cachedValues;
+
+  public ColumnProfile(String name, ColumnSchema schema, int cacheSize) {
+    this.name = name;
+    this.schema = schema;
+    this.cacheSize = cacheSize; // TODO: add logger for size of cache on column
+
+    // Defaulted
+    this.metrics = this.schema.getMetrics();
+    this.projector = new SingleFieldProjector<>(name);
+    this.successCount = 0;
+    this.failureCount = 0;
+    this.nullCount = 0;
+    this.cachedValues = new ArrayList<>();
+  }
+
+  public void addMetric(Metric<?> metric) {
+    if (this.metrics.containsKey(metric.getNamespace())) {
+      // TODO: Add logger with warning about replacement
+    }
+
+    this.metrics.put(metric.getNamespace(), metric);
+    // TODO: Wouldn't this implement a success count here?
+  }
+
+  // TODO: this only gets one not every part of the row. Should projector actually do it multiple?
+  public void track(HashMap<String, Object> row) {
+    T value = this.projector.apply(row);
+    this.cachedValues.add(value);
+
+    if (this.cachedValues.size() >= this.cacheSize) {
+      this.flush();
+    }
+  }
+
+  public void flush() {
+    // TODO: Logger was initially here, but only for when it was forced, think it through
+    ArrayList<T> oldCache = this.cachedValues;
+    this.cachedValues = new ArrayList<>();
```

Comment:
How expensive is allocation/littering in Java? Would it be worthwhile to maintain an index into the cache so we can re-use it? (ok as a follow-up PR if we decide to do it).

Reply:
Good point. Yes let's go ahead just cause that is a little better, but let's do it in another PR